### PR TITLE
.Net: Migrate Azure Chat Completion Service to AzureOpenAI SDK v2

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
### Motivation and Context
This PR is the next step in a series of follow-up PRs to migrate AzureOpenAIConnector to Azure AI SDK v2. It updates all code related to AzureOpenAI ChatCompletionService to use the Azure AI SDK v2. One of the goals of the PR is to update the code with a minimal number of changes to make the code review as easy as possible, so almost all methods keep their names as they were even though they might not be relevant anymore. This will be fixed in one of the follow-up PRs.

### Description
This PR does the following:  
1. Migrates AzureOpenAIChatCompletionService, ClientCore, and other model classes both use, to Azure AI SDK v2.  
2. Updates ToolCallBehavior classes to return a list of functions and function choice. This change is required because the new SDK model requires both of those for the CompletionsOptions class creation and does not allow setting them after the class is already created, as it used to allow.  
3. Adapts related unit tests to the API changes.

### Next steps
1. Add integration tests.  
2. Rename internal/private methods that were intentionally left with old, irrelevant names to minimize the code review delta.

### Out of scope:
* https://github.com/microsoft/semantic-kernel/issues/6991

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 1]
```diff
@@ -1132,52 +861,51 @@ private static CompletionsOptions CreateCompletionsOptions(string text, AzureOpe
...
+    }
+
+    private static List<ChatMessage> CreateChatCompletionMessages(AzureOpenAIPromptExecutionSettings executionSettings, ChatHistory chatHistory)
+    {
+        List<ChatMessage> messages = [];
+
         if (!string.IsNullOrWhiteSpace(executionSettings.ChatSystemPrompt) && !chatHistory.Any(m => m.Role == AuthorRole.System))
         {
-            options.Messages.AddRange(GetRequestMessages(new ChatMessageContent(AuthorRole.System, executionSettings!.ChatSystemPrompt), executionSettings.ToolCallBehavior));
+            messages.AddRange(GetRequestMessages(new ChatMessageContent(AuthorRole.System, executionSettings!.ChatSystemPrompt), executionSettings.ToolCallBehavior));
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:873]: Do we need to create a message content to get the request?

```suggestion
            messages.AddRange(new SystemChatMessage(executionSettings!.ChatSystemPrompt));
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 2]
```diff
@@ -1292,69 +1020,60 @@ private static List<ChatRequestMessage> GetRequestMessages(ChatMessageContent me
...
-            return new ChatRequestUserMessage(message.Content);
+            return ChatMessage.CreateUserMessage(completion.Content);
         }
 
-        throw new NotSupportedException($"Role {message.Role} is not supported.");
+        throw new NotSupportedException($"Role {completion.Role} is not supported.");
     }
 
-    private AzureOpenAIChatMessageContent GetChatMessage(ChatChoice chatChoice, ChatCompletions responseData)
+    private AzureOpenAIChatMessageContent GetChatMessage(OpenAI.Chat.ChatCompletion completion)
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]:
```suggestion
    private AzureOpenAIChatMessageContent GetChatMessage(OpenAIChatCompletion completion)
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 3]
```diff
@@ -1369,21 +1088,21 @@ private AzureOpenAIChatMessageContent GetChatMessage(ChatRole chatRole, string c
         return message;
     }
 
-    private IEnumerable<FunctionCallContent> GetFunctionCallContents(IEnumerable<ChatCompletionsToolCall> toolCalls)
+    private IEnumerable<FunctionCallContent> GetFunctionCallContents(IEnumerable<ChatToolCall> toolCalls)
     {
         List<FunctionCallContent>? result = null;
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:1093]: nit.

```suggestion
        List<FunctionCallContent> result = [];
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 4]
```diff
@@ -1424,7 +1143,7 @@ private IEnumerable<FunctionCallContent> GetFunctionCallContents(IEnumerable<Cha
         return result ?? Enumerable.Empty<FunctionCallContent>();
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:1143]: nit.

```suggestion
        return result;
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 5]
```diff
@@ -1459,21 +1178,23 @@ private static void ValidateMaxTokens(int? maxTokens)
...
 
-    private static async Task<T> RunRequestAsync<T>(Func<Task<T>> request)
+    private static T RunRequest<T>(Func<T> request)
     {
         try
         {
-            return await request.Invoke().ConfigureAwait(false);
+            return request.Invoke();
         }
         catch (RequestFailedException e)
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:1199]: New OpenAI client throws a `System.ClientModel`, `ClientResultException`, check for it.

```suggestion
        catch (ClientResultException cre)
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 6]
```diff
@@ -1571,4 +1292,95 @@ private void LogUsage(CompletionsUsage usage)
...
+            }
+        }
+
+        return new ToolCallingConfig(
+            Tools: tools ?? [s_nonInvocableFunctionTool],
+            Choice: choice ?? ChatToolChoice.None,
+            AutoInvoke: autoInvoke);
+    }
+
+    private static ChatResponseFormat? GetResponseFormat(AzureOpenAIPromptExecutionSettings executionSettings)
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:1336]: nit.

Might consider using switch expressions to reduce the verbosity.

```suggestion
    private static ChatResponseFormat? GetResponseFormat(AzureOpenAIPromptExecutionSettings executionSettings)
```

File: dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs [hunk 7]
```diff
@@ -133,22 +132,27 @@ internal sealed class KernelFunctions : AzureToolCallBehavior
...
                 // Provide all functions from the kernel.
                 IList<KernelFunctionMetadata> functions = kernel.Plugins.GetFunctionsMetadata();
                 if (functions.Count > 0)
                 {
-                    options.ToolChoice = ChatCompletionsToolChoice.Auto;
+                    choice = ChatToolChoice.Auto;
                     for (int i = 0; i < functions.Count; i++)
                     {
-                        options.Tools.Add(new ChatCompletionsFunctionToolDefinition(functions[i].ToAzureOpenAIFunction().ToFunctionDefinition()));
+                        (tools ??= []).Add(functions[i].ToAzureOpenAIFunction().ToFunctionDefinition());
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs:150]: In theory we can, but should we? Taking into account that the functionality has been there for ages, and we are going to replace it by the new function calling model, I would not do that, at least now.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIChatMessageContent.cs [hunk 8]
```diff
@@ -13,28 +13,28 @@ namespace Microsoft.SemanticKernel.Connectors.AzureOpenAI;
 public sealed class AzureOpenAIChatMessageContent : ChatMessageContent
 {
     /// <summary>
-    /// Gets the metadata key for the <see cref="ChatCompletionsToolCall.Id"/> name property.
+    /// Gets the metadata key for the tool id.
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIChatMessageContent.cs:16]: This property represents the metadata key for the function call id. The client core attempts to find the key in chat history messages to determine if any of them represent a function result, which is indicated by the Content property of type string. Therefore, this property is the id for the function result, not the function call. In any case, this property, along with the nearby one, is likely to be removed soon.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 9]
```diff
@@ -30,8 +34,12 @@ namespace Microsoft.SemanticKernel.Connectors.AzureOpenAI;
 /// </summary>
 internal abstract class ClientCore
 {
+    private const string CreatedMetadataKey = "Created";
+    private const string PromptFilterResultsMetadataKey = "PromptFilterResults";
+    private const string ContentFilterResultsMetadataKey = "ContentFilterResults";
+    private const string LogProbabilityInfoMetadataKey = "LogProbabilityInfo";
     private const string ModelProvider = "openai";
-    private const int MaxResultsPerPrompt = 128;
+    private record ToolCallingConfig(IList<ChatTool>? Tools, ChatToolChoice Choice, bool AutoInvoke);
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:42]: There's no particular reason; it's just to be simply declared in one line. Context: All tool call behavior functionality is going to be removed, so it's not really important at the moment.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 10]
```diff
@@ -382,30 +254,36 @@ internal ClientCore(ILogger? logger = null)
...
 
+        if (this.Logger.IsEnabled(LogLevel.Trace))
+        {
+            this.Logger.LogTrace("ChatHistory: {ChatHistory}, Settings: {Settings}",
+                JsonSerializer.Serialize(chat),
+                JsonSerializer.Serialize(executionSettings));
+        }
+
         // Convert the incoming execution settings to OpenAI settings.
         AzureOpenAIPromptExecutionSettings chatExecutionSettings = AzureOpenAIPromptExecutionSettings.FromExecutionSettings(executionSettings);
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:265]: Yep, put it in my todo list. Will update it in a scope of a different PR.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs [hunk 11]
```diff
@@ -382,30 +254,36 @@ internal ClientCore(ILogger? logger = null)
...
-        bool autoInvoke = kernel is not null && chatExecutionSettings.ToolCallBehavior?.MaximumAutoInvokeAttempts > 0 && s_inflightAutoInvokes.Value < MaxInflightAutoInvokes;
+
         ValidateMaxTokens(chatExecutionSettings.MaxTokens);
-        ValidateAutoInvoke(autoInvoke, chatExecutionSettings.ResultsPerPrompt);
 
-        // Create the Azure SDK ChatCompletionOptions instance from all available information.
-        var chatOptions = this.CreateChatCompletionsOptions(chatExecutionSettings, chat, kernel, this.DeploymentOrModelName);
+        var chatMessages = CreateChatCompletionMessages(chatExecutionSettings, chat);
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:269]: It was done on purpose, as discussed earlier. Renaming will go as a separate PR.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 6 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs:1336]: I might do it some other time, there are more important things to finish.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 5 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]: Yep, thanks, fixed.

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 4 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 3 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs (hunk 1 above)
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/ClientCore.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs [hunk 12]
```diff
@@ -211,29 +218,31 @@ internal override void ConfigureOptions(Kernel? kernel, ChatCompletionsOptions o
                     }
 
                     // Add the function.
-                    options.Tools.Add(functions[i]);
+                    (tools ??= []).Add(functions[i]);
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIChatMessageContent.cs [hunk 13]
```diff
@@ -13,28 +13,28 @@ namespace Microsoft.SemanticKernel.Connectors.AzureOpenAI;
...
     /// </summary>
-    internal static string FunctionToolCallsProperty => $"{nameof(ChatResponseMessage)}.FunctionToolCalls";
+    internal static string FunctionToolCallsProperty => "ChatResponseMessage.FunctionToolCalls";
 
     /// <summary>
     /// Initializes a new instance of the <see cref="AzureOpenAIChatMessageContent"/> class.
     /// </summary>
-    internal AzureOpenAIChatMessageContent(ChatResponseMessage chatMessage, string modelId, IReadOnlyDictionary<string, object?>? metadata = null)
-        : base(new AuthorRole(chatMessage.Role.ToString()), chatMessage.Content, modelId, chatMessage, System.Text.Encoding.UTF8, CreateMetadataDictionary(chatMessage.ToolCalls, metadata))
+    internal AzureOpenAIChatMessageContent(OpenAI.Chat.ChatCompletion completion, string modelId, IReadOnlyDictionary<string, object?>? metadata = null)
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIChatMessageContent.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs [hunk 14]
```diff
@@ -133,22 +132,27 @@ internal sealed class KernelFunctions : AzureToolCallBehavior
...
+
             // If no kernel is provided, we don't have any tools to provide.
             if (kernel is not null)
             {
                 // Provide all functions from the kernel.
                 IList<KernelFunctionMetadata> functions = kernel.Plugins.GetFunctionsMetadata();
                 if (functions.Count > 0)
                 {
-                    options.ToolChoice = ChatCompletionsToolChoice.Auto;
+                    choice = ChatToolChoice.Auto;
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs:147]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/AzureToolCallBehavior.cs:1]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI.UnitTests/AzureOpenAITestHelper.cs [hunk 15]
```diff
@@ -17,4 +19,13 @@ internal static string GetTestResponse(string fileName)
...
         return File.ReadAllText($"./TestData/{fileName}");
     }
+
+    /// <summary>
+    /// Reads test response from file and create <see cref="StreamContent"/>.
+    /// </summary>
+    /// <param name="fileName">Name of the file with test response.</param>
+    internal static StreamContent GetTestResponseAsStream(string fileName)
+    {
+        return new StreamContent(new MemoryStream(Encoding.UTF8.GetBytes(AzureOpenAITestHelper.GetTestResponse(fileName))));
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI.UnitTests/AzureOpenAITestHelper.cs]: fixed, thanks!

File: dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIStreamingChatMessageContent.cs [hunk 16]
```diff
@@ -77,11 +78,27 @@ public sealed class AzureOpenAIStreamingChatMessageContent : StreamingChatMessag
...
     /// <inheritdoc/>
     public override string ToString() => this.Content ?? string.Empty;
+
+    private static StreamingKernelContentItemCollection CreateContentItems(IReadOnlyList<ChatMessageContentPart> contentUpdate)
+    {
+        StreamingKernelContentItemCollection collection = [];
+
+        foreach (var content in contentUpdate)
+        {
+            // We only support text content for now.
```
[the remark this answers is not in the document]
Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.AzureOpenAI/Core/AzureOpenAIStreamingChatMessageContent.cs:95]: Yep, done.

**Review by RogerBarreto — COMMENTED** [review]
Approved. ✅

**Pull request merged**